File: algorithm/disjoint_set.cc
#include <bits/stdc++.h>

using namespace std;

class DisjointSet {
 public:
  DisjointSet(int n) : parent_(n), rank_(n) {
    iota(parent_.begin(), parent_.end(), 0);
  }

  int find(int x) {
    if (x != parent_[x]) {
      parent_[x] = find(parent_[x]);
    }
    return parent_[x];
  }

  void join(int x, int y) {
    int px = find(x), py = find(y);
    if (px == py) return;
    if (rank_[px] > rank_[py]) {
      parent_[py] = px;
    } else if (rank_[py] > rank_[px]) {
      parent_[px] = py;
    } else {
      parent_[py] = px;
      rank_[px]++;
    }
  }

 private:
  vector<int> parent_, rank_;
};

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  DisjointSet ds(10);
  ds.join(0, 1);
  ds.join(1, 3);
  ds.join(2, 5);
  ds.join(2, 6);

  unordered_map<int, vector<int>> groups;
  for (int i = 0; i < 10; ++i) {
    groups[ds.find(i)].push_back(i);
  }
  for (auto g : groups) {
    cout << "group " << g.first << " has : ";
    for (int item : g.second) {
      cout << item << " ";
    }
    cout << endl;
  }
  return 0;
}

File: algorithm/disjoint_set.rs
struct DisjointSet {
    parent: Vec<usize>,
    rank: Vec<i32>,
}

impl DisjointSet {
    fn new(n: usize) -> Self {
        DisjointSet {
            parent: (0..n).collect(),
            rank: vec![0; n],
        }
    }

    fn find(&mut self, x: usize) -> usize {
        if x != self.parent[x] {
            self.parent[x] = self.find(self.parent[x]);
        }
        self.parent[x]
    }

    fn join(&mut self, x: usize, y: usize) {
        let px = self.find(x);
        let py = self.find(y);
        if px == py {
            return;
        }
        if self.rank[px] > self.rank[py] {
            self.parent[py] = px;
        } else if self.rank[py] > self.rank[px] {
            self.parent[px] = py;
        } else {
            self.parent[py] = px;
            self.rank[px] += 1;
        }
    }
}

fn main() {
    let mut ds = DisjointSet::new(10);
    ds.join(0, 1);
    ds.join(1, 3);
    ds.join(2, 5);
    ds.join(2, 6);

    use std::collections::HashMap;
    let mut groups: HashMap<usize, Vec<usize>> = HashMap::new();
    for i in 0..10 {
        let root = ds.find(i);
        groups.entry(root).or_insert(Vec::new()).push(i);
    }

    for (group_id, members) in groups {
        print!("group {} has : ", group_id);
        for &member in &members {
            print!("{} ", member);
        }
        println!();
    }
}
